### PR TITLE
 Dealing with categorical and integer-valued variables in Bayesian Optimization with Gaussian processes

### PR DESCRIPTION
**Pull request by Paul-Saves**
Hello,
This is the mixed ego optimization with mixed models.

How to use?
By default, the type of the variables is assumed continuous.
If not, you can pass a mixed surrogate s.

s = KRG( vartype=["cont","int",("cate",3),("cate",2)])
xlimits= np.array([ [-1,2.5],[0,2], ["square","circle",rhombus"],["green","blue" ]])
ego = EGO(n_doe=n_doe,n_iter=n_iter, criterion=criterion,xlimits=xlimits,qEI=qEI,tunnel=0, surrogate=s )

You can find it in the new documentation.

I have written here the pseudo-code:
image
Reference: [GMHL.pdf](https://github.com/SMTorg/smt/files/4967282/GMHL.pdf)

### CODE REVIEW

**Comment by Paul-Saves** [comment]
Add assign_labels to outputs of ego: x_data

**Comment by Paul-Saves** [comment]
h

File: doc/_src_docs/applications/ego.rstx [hunk 1]
```diff
@@ -159,6 +159,10 @@ Beside the Expected Improvement, the implementation here offers two other infill
 Regarding the parallel execution, one can implement specific multiprocessing by deriving the _Evaluator_ interface
 and overriding the default implementation of the _run(fun, x)_ method. The default implementation simply runs _fun(x)_.
 
+Regardless the others parameters, you can specify a mixed surrogate model to make mixed optimization. [8]_
+
+You can penalize already evaluated point in EI via tunneling (not recommanded for high dimensional problems).
```
Comment (relf) [inline doc/_src_docs/applications/ego.rstx:164]: Maybe add a sentence or two to explain the tunneling principle a bit more... (why I would do that? what is it?)

File: smt/applications/ego.py [hunk 2]
```diff
@@ -99,6 +99,8 @@ def _initialize(self):
         declare("ydoe", None, types=np.ndarray, desc="Initial doe outputs")
         declare("xlimits", None, types=np.ndarray, desc="Bounds of function fun inputs")
         declare("verbose", False, types=bool, desc="Print computation information")
+        declare("tunnel", 0, types=int, desc="1 to enable tunneling in ei")
```
Comment (relf) [inline smt/applications/ego.py:102]: tunnel should be better implemented by a boolean (see verbose above) and maybe renamed enable_tunneling with further explanations in the description (enable the penalization of points that have been already evaluated in EI criterion).

File: smt/applications/ego.py [hunk 3]
```diff
@@ -99,6 +99,8 @@ def _initialize(self):
...
+        declare("surrogate", KRG(print_global=False), desc="surrogate model to use")
```
Comment (relf) [inline smt/applications/ego.py:103]: I think you can declare types=KrgBased or more strictly types=(KRG, KPLS, KPLSK) maybe description could be more specific "SMT kriging-based surrogate model used internally"

File: smt/applications/tests/test_ego.py [hunk 4]
```diff
@@ -47,13 +53,15 @@ def function_test_1d(x):
         return y.reshape((-1, 1))
 
     def test_evaluator(self):
+        print("evaluator")
```
Comment (relf) [inline smt/applications/tests/test_ego.py:56]: print information given by the test runner, could be removed

File: smt/applications/tests/test_ego.py [hunk 5]
```diff
@@ -35,6 +38,9 @@ def run(self, fun, x):
 
 
 class TestEGO(SMTestCase):
+    """
+    Test class
```
Comment (relf) [inline smt/applications/tests/test_ego.py:42]: Well... less informative than the name of the class above, could be removed

File: smt/applications/ego.py [hunk 6]
```diff
@@ -207,33 +226,38 @@ def _setup_optimizer(self, fun):
...
         else:
             self.log("Initial DOE given")
             x_doe = np.atleast_2d(xdoe)
 
+        x_doe = self.gpr._project_values(x_doe)
         ydoe = self.options["ydoe"]
         if ydoe is None:
-            y_doe = self._evaluator.run(fun, x_doe)
+            y_doe = self._evaluator.run(
+                fun, self.gpr._assign_labels(x_doe, self.options["xlimits"])
```
Comment (relf) [inline smt/applications/ego.py:253]: _assign_labels should be renamed assign_labels. The underscore in front of a method name specify a protected method normally not used outside of the class or subclasses. same thing for project_values

**Review by relf — CHANGES_REQUESTED** [review]

**Review by relf — APPROVED** [review]
Thanks for this new feature in SMT!

**Comment by Paul-Saves** [comment]
Thanks a lot for your help !

**Pull request merged**